hermes_core/util/cdf.py

```diff
@@ -0,0 +1,612 @@
+"""
+This module provides CDF file functionality.
+
+"""
+import datetime
+from pathlib import Path
+import yaml
+from spacepy import pycdf
+import numpy as np
+import hermes_core
+from hermes_core import log
+from hermes_core.util import util
+from hermes_core.util.exceptions import warn_user
+
+DEFAULT_SCHEMA_FILE = "hermes_default_schema.yaml"
+DEFAULT_ATTRIBUTES_FILE = "hermes_default_attributes.yaml"
+
+
+class CDFWriter:
+    """
+    Python Class to Convert abstract Data Structures to CDF file
+    formats compliant with SpacePy's pyCDF Module. These CDF Files
+    can then be indexed and shared through SPDF.
+
+    The purpose of this class is to provide an intermediate interafce to generate CDF Files.
+    One limitation of the pyCDF library is that you can not create CDF-like data structures
+    in memory without generating and saving a local `.cdf` File. This class provides an
+    interface to manipulate CDF-like data structures in memory before the generation and saving
+    of a local `.cdf` file.
+
+    This class uses an internal native python `dict` data structure to provide a CDF-like
+    data structure that can be manipulated in memory. The internal `dict` used `self.global_attrs`
+    is formatted in the following way.
+
+    ```py
+    self.global_attrs = {
+        'global_attribute_name1': 'global_attribute_value1',
+        'global_attribute_name2': 'global_attribute_value2',
+    }
+
+    ```
+    """
+
+    def __init__(self):
+
+        # Current Working CDF File
+        self.target_cdf_file = None
+
+        # Data Validation, Complaiance, Derived Attributes
+        self.schema = {}
+        # Load Default schema Data
+        self._load_default_schema_data()
+
+        # Intermediate Data Structure for Global / Metadata Attributes
+        self.global_attrs = {}
+        # Load Default Global Attributes
+        self._load_default_attributes()
+
+        # Intermedaite Data Structure for Variable Data
+        self.variable_data = {}
```

Comment:
It is a standard in a number of scientific python packages to use `.meta` and `.data`. Let's use .meta for the global attrs, and then attach the instrument meta data to the .data attribute. We could consider using the [NDData](https://docs.astropy.org/en/stable/nddata/index.html) object for each data object. The attribute .data would then be a list of NDData object each with its own .meta.

Reply:
Are you suggesting `a.data` be a dict of `(str, NDData)` pairs? I was looking at the NDData class and theres no way to contain a name for the variable other than in the contained `.meta` dict. 
Are you thinking:
```py
a = CDFWriter()
...
a..data -> {}
a.data['epoch'] -> NDData
```
or
```py
a = CDFWriter()
...
a.data -> list(NDData)
a.data[0].data -> np.ndarray
data[0].meta -> {
                                  'name':'epoch',
                                  ....
                             }
```

Reply:
As we discussed, a.data should be an astropy TimeSeries object.